_posts/2016-04-26-aws-and-consul.md

```diff
@@ -0,0 +1,175 @@
+
+```
+# Create an ELB load balancer to sit in front
+resource "aws_elb" "consul" {
+    name = "consul"
+    subnets = ["s-abc123"]
+    security_groups = ["sg-abc123"]
+
+    listener {
+        instance_port = 8500
+        instance_protocol = "http"
+        lb_port = 80
+        lb_protocol = "http"
+    }
+
+    health_check {
+        healthy_threshold = 2
+        unhealthy_threshold = 2
+        timeout = 3
+        target = "HTTP:8500/v1/status/leader"
+        interval = 30
+    }
+    
+    lifecycle {
+        create_before_destroy = true
+    }
+}
+
+# Define the ASG
+resource "aws_autoscaling_group" "consul" {
+    name = "consul - ${aws_launch_configuration.consul.name}"
+    launch_configuration = "${aws_launch_configuration.consul.name}"
+    desired_capacity = 5
+    min_size = 3
+    max_size = 5
+    min_elb_capacity = 3
+    load_balancers = ["${aws_elb.consul.id}"]
+
+    lifecycle {
+        create_before_destroy = true
+    }
+    
+    tag {
+        key = "Name"
+        value = "consul"
+        propagate_at_launch = true
+    }
+}
+
+# Define it's Launch Configuration
+resource "aws_launch_configuration" "consul" {
+    image_id  = "ami-abc123"
+    instance_type = "t2.medium"
+    key_name = "deploy-key"
+    security_groups = ["sg-abc123"]
+
+    # Make sure our instance has a policy which grants read-only access to EC2 API
+    iam_instance_profile = "arn:aws:iam::0123456789:some_profile/ec2"
+
+    # More on user_data later on
+    user_data = "${template_file.consul_userdata.rendered}"
+
+    root_block_device {
+        volume_type = "gp2"
+        volume_size = 10
+    }
+
+    lifecycle {
+        create_before_destroy = true
+    }
+}
+```
+**Nb.** This is simplified a lot, in a real example, much would be variable-ised, but for the sake of a self-contained example I have used real (made-up) values instead of variables.
+
+Nothing too crazy here, we have an ASG + LC which will launch 5 instances based off our AMI. The cleverness here, which is a bit specific to Terraform but can be replicated in other ways using Cloud Formation for example, is the `create_before_destroy = true` statement. So what does that mean? If you change the AMI id, it will create a new ASG+LC alongside the old one. Only when the new instances are attached to the ELB (Elastic Load Balancer) and the ELB declares them healthy will the old instances be torn down. 
+
+### User Data
+If we build this now, we will get a set of 5 Consul nodes, we won't get a Cluster. As I alluded to before, we need to do a little post-launch config to get the nodes to start talking to each other. If we make sure our Consul AMI has the AWS CLI tools installed, its fairly easy to use that to query AWS and ask for all instances tagged 'consul' and get their IP addresses and then shove those values into Consul's config file and finally, start the service.
```

Comment:
Again, "five" instead of "5", "It's fairly easy", not "Its fairly easy".  Remove the "and" before "get their IP addresses" and and replace it with a comma and, similarly, replace the "and then" before "shove those values" with a comma.